<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Using the GEOquery Package &bull; GEOquery</title><!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous"><script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous"><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <div class="container">
      <header><div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">GEOquery</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/seandavi/GEOquery">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Using the GEOquery Package</h1>
                        <h4 class="author">Sean Davis</h4>
            
            <h4 class="date">September 21, 2014</h4>
          </div>

    
    
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Using GEOquery}
-->
<div id="overview-of-geo" class="section level1">
<h1>Overview of GEO</h1>
<p>The NCBI Gene Expression Omnibus (GEO) serves as a public repository for a wide range of high-throughput experimental data. These data include single and dual channel microarray-based experiments measuring mRNA, genomic DNA, and protein abundance, as well as non-array techniques such as serial analysis of gene expression (SAGE), mass spectrometry proteomic data, and high-throughput sequencing data.</p>
<p>At the most basic level of organization of GEO, there are four basic entity types. The first three (Sample, Platform, and Series) are supplied by users; the fourth, the dataset, is compiled and curated by GEO staff from the user-submitted data. See <a href="http://www.ncbi.nih.gov/geo">the GEO home page</a> for more information.</p>
<div id="platforms" class="section level2">
<h2>Platforms</h2>
<p>A Platform record describes the list of elements on the array (e.g., cDNAs, oligonucleotide probesets, ORFs, antibodies) or the list of elements that may be detected and quantified in that experiment (e.g., SAGE tags, peptides). Each Platform record is assigned a unique and stable GEO accession number (GPLxxx). A Platform may reference many Samples that have been submitted by multiple submitters.</p>
</div>
<div id="samples" class="section level2">
<h2>Samples</h2>
<p>A Sample record describes the conditions under which an individual Sample was handled, the manipulations it underwent, and the abundance measurement of each element derived from it. Each Sample record is assigned a unique and stable GEO accession number (GSMxxx). A Sample entity must reference only one Platform and may be included in multiple Series.</p>
</div>
<div id="series" class="section level2">
<h2>Series</h2>
<p>A Series record defines a set of related Samples considered to be part of a group, how the Samples are related, and if and how they are ordered. A Series provides a focal point and description of the experiment as a whole. Series records may also contain tables describing extracted data, summary conclusions, or analyses. Each Series record is assigned a unique and stable GEO accession number (GSExxx). Series records are available in a couple of formats which are handled by GEOquery independently. The smaller and new GSEMatrix files are quite fast to parse; a simple flag is used by GEOquery to choose to use GSEMatrix files (see below).</p>
</div>
<div id="datasets" class="section level2">
<h2>Datasets</h2>
<p>GEO DataSets (GDSxxx) are curated sets of GEO Sample data. A GDS record represents a collection of biologically and statistically comparable GEO Samples and forms the basis of GEO&rsquo;s suite of data display and analysis tools. Samples within a GDS refer to the same Platform, that is, they share a common set of probe elements. Value measurements for each Sample within a GDS are assumed to be calculated in an equivalent manner, that is, considerations such as background processing and normalization are consistent across the dataset. Information reflecting experimental design is provided through GDS subsets.</p>
</div>
</div>
<div id="getting-started-using-geoquery" class="section level1">
<h1>Getting Started using GEOquery</h1>
<p>Getting data from GEO is really quite easy. There is only one command that is needed, <code>getGEO</code>. This one function interprets its input to determine how to get the data from GEO and then parse the data into useful R data structures. Usage is quite simple. This loads the GEOquery library.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(GEOquery)</code></pre></div>
<p>Now, we are free to access any GEO accession. <em>Note that in the following, I use a file packaged with the GEOquery package. In general, you will use only the GEO accession, as noted in the code comments.</em></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># If you have network access, the more typical way to do this</span>
<span class="co"># would be to use this:</span>
<span class="co"># gds &lt;- getGEO("GDS507")</span>
gds &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getGEO.html">getGEO</a></span>(<span class="dt">filename=</span><span class="kw">system.file</span>(<span class="st">"extdata/GDS507.soft.gz"</span>,<span class="dt">package=</span><span class="st">"GEOquery"</span>))</code></pre></div>
<p>Now, <code>gds</code> contains the R data structure (of class <code>GDS</code>) that represents the GDS507 entry from GEO. You&rsquo;ll note that the filename used to store the download was output to the screen (but not saved anywhere) for later use to a call to <code><a href="../reference/getGEO.html">getGEO(filename=...)</a></code>.</p>
<p>We can do the same with any other GEO accession, such as <code>GSM11805</code>, a GEO sample.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># If you have network access, the more typical way to do this</span>
<span class="co"># would be to use this:</span>
<span class="co"># gds &lt;- getGEO("GSM11805")</span>
gsm &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getGEO.html">getGEO</a></span>(<span class="dt">filename=</span><span class="kw">system.file</span>(<span class="st">"extdata/GSM11805.txt.gz"</span>,<span class="dt">package=</span><span class="st">"GEOquery"</span>))</code></pre></div>
</div>
<div id="geoquery-data-structures" class="section level1">
<h1>GEOquery Data Structures</h1>
<p>The GEOquery data structures really come in two forms. The first, comprising <code>GDS</code>, <code>GPL</code>, and <code>GSM</code> all behave similarly and accessors have similar effects on each. The fourth GEOquery data structure, <code>GSE</code> is a composite data type made up of a combination of <code>GSM</code> and <code>GPL</code> objects. I will explain the first three together first.</p>
<div id="the-gds-gsm-and-gpl-classes" class="section level2">
<h2>The GDS, GSM, and GPL classes</h2>
<p>Each of these classes is comprised of a metadata header (taken nearly verbatim from the SOFT format header) and a GEODataTable. The GEODataTable has two simple parts, a Columns part which describes the column headers on the Table part. There is also a <code>show</code> method for each class. For example, using the gsm from above:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Look at gsm metadata:</span>
<span class="kw">head</span>(<span class="kw"><a href="../reference/Generic%20functions.html">Meta</a></span>(gsm))</code></pre></div>
<pre><code>## $channel_count
## [1] "1"
## 
## $comment
## [1] "Raw data provided as supplementary file"
## 
## $contact_address
## [1] "715 Albany Street, E613B"
## 
## $contact_city
## [1] "Boston"
## 
## $contact_country
## [1] "USA"
## 
## $contact_department
## [1] "Genetics and Genomics"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Look at data associated with the GSM:</span>
<span class="co"># but restrict to only first 5 rows, for brevity</span>
<span class="kw"><a href="../reference/Generic%20functions.html">Table</a></span>(gsm)[<span class="dv">1</span>:<span class="dv">5</span>,]</code></pre></div>
<pre><code>##           ID_REF  VALUE ABS_CALL
## 1 AFFX-BioB-5_at  953.9        P
## 2 AFFX-BioB-M_at 2982.8        P
## 3 AFFX-BioB-3_at 1657.9        P
## 4 AFFX-BioC-5_at 2652.7        P
## 5 AFFX-BioC-3_at 2019.5        P</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Look at Column descriptions:</span>
<span class="kw"><a href="../reference/Generic%20functions.html">Columns</a></span>(gsm)</code></pre></div>
<pre><code>##     Column
## 1   ID_REF
## 2    VALUE
## 3 ABS_CALL
##                                                                  Description
## 1                                                                           
## 2                         MAS 5.0 Statistical Algorithm (mean scaled to 500)
## 3 MAS 5.0 Absent, Marginal, Present call  with Alpha1 = 0.05, Alpha2 = 0.065</code></pre>
<p>The <code>GPL</code> class behaves exactly as the <code>GSM</code> class. However, the <code>GDS</code> class has a bit more information associated with the <code>Columns</code> method:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/Generic%20functions.html">Columns</a></span>(gds)[,<span class="dv">1</span>:<span class="dv">3</span>]</code></pre></div>
<pre><code>##      sample disease.state individual
## 1  GSM11815           RCC        035
## 2  GSM11832           RCC        023
## 3  GSM12069           RCC        001
## 4  GSM12083           RCC        005
## 5  GSM12101           RCC        011
## 6  GSM12106           RCC        032
## 7  GSM12274           RCC          2
## 8  GSM12299           RCC          3
## 9  GSM12412           RCC          4
## 10 GSM11810        normal        035
## 11 GSM11827        normal        023
## 12 GSM12078        normal        001
## 13 GSM12099        normal        005
## 14 GSM12269        normal          1
## 15 GSM12287        normal          2
## 16 GSM12301        normal          3
## 17 GSM12448        normal          4</code></pre>
</div>
<div id="the-gse-class" class="section level2">
<h2>The GSE class</h2>
<p>The <code>GSE</code> entity is the most confusing of the GEO entities. A GSE entry can represent an arbitrary number of samples run on an arbitrary number of platforms. The <code>GSE</code> class has a metadata section, just like the other classes. However, it doesn&rsquo;t have a GEODataTable. Instead, it contains two lists, accessible using the <code>GPLList</code> and <code>GSMList</code> methods, that are each lists of <code>GPL</code> and <code>GSM</code> objects. To show an example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Again, with good network access, one would do:</span>
<span class="co"># gse &lt;- getGEO("GSE781",GSEMatrix=FALSE)</span>
gse &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getGEO.html">getGEO</a></span>(<span class="dt">filename=</span><span class="kw">system.file</span>(<span class="st">"extdata/GSE781_family.soft.gz"</span>,<span class="dt">package=</span><span class="st">"GEOquery"</span>))</code></pre></div>
<pre><code>## Parsing....</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw"><a href="../reference/Generic%20functions.html">Meta</a></span>(gse))</code></pre></div>
<pre><code>## $contact_address
## [1] "715 Albany Street, E613B"
## 
## $contact_city
## [1] "Boston"
## 
## $contact_country
## [1] "USA"
## 
## $contact_department
## [1] "Genetics and Genomics"
## 
## $contact_email
## [1] "mlenburg@bu.edu"
## 
## $contact_fax
## [1] "617-414-1646"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># names of all the GSM objects contained in the GSE</span>
<span class="kw">names</span>(<span class="kw"><a href="../reference/Generic%20functions.html">GSMList</a></span>(gse))</code></pre></div>
<pre><code>##  [1] "GSM11805" "GSM11810" "GSM11814" "GSM11815" "GSM11823" "GSM11827"
##  [7] "GSM11830" "GSM11832" "GSM12067" "GSM12069" "GSM12075" "GSM12078"
## [13] "GSM12079" "GSM12083" "GSM12098" "GSM12099" "GSM12100" "GSM12101"
## [19] "GSM12105" "GSM12106" "GSM12268" "GSM12269" "GSM12270" "GSM12274"
## [25] "GSM12283" "GSM12287" "GSM12298" "GSM12299" "GSM12300" "GSM12301"
## [31] "GSM12399" "GSM12412" "GSM12444" "GSM12448"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># and get the first GSM object on the list</span>
<span class="kw"><a href="../reference/Generic%20functions.html">GSMList</a></span>(gse)[[<span class="dv">1</span>]]</code></pre></div>
<pre><code>## An object of class "GSM"
## channel_count 
## [1] "1"
## comment 
## [1] "Raw data provided as supplementary file"
## contact_address 
## [1] "715 Albany Street, E613B"
## contact_city 
## [1] "Boston"
## contact_country 
## [1] "USA"
## contact_department 
## [1] "Genetics and Genomics"
## contact_email 
## [1] "mlenburg@bu.edu"
## contact_fax 
## [1] "617-414-1646"
## contact_institute 
## [1] "Boston University School of Medicine"
## contact_name 
## [1] "Marc,E.,Lenburg"
## contact_phone 
## [1] "617-414-1375"
## contact_state 
## [1] "MA"
## contact_web_link 
## [1] "http://gg.bu.edu"
## contact_zip/postal_code 
## [1] "02130"
## data_row_count 
## [1] "22283"
## description 
## [1] "Age = 70; Gender = Female; Right Kidney; Adjacent Tumor Type = clear cell; Adjacent Tumor Fuhrman Grade = 3; Adjacent Tumor Capsule Penetration = true; Adjacent Tumor Perinephric Fat Invasion = true; Adjacent Tumor Renal Sinus Invasion = false; Adjacent Tumor Renal Vein Invasion = true; Scaling Target = 500; Scaling Factor = 7.09; Raw Q = 2.39; Noise = 2.60; Background = 55.24."
## [2] "Keywords = kidney"                                                                                                                                                                                                                                                                                                                                                                           
## [3] "Keywords = renal"                                                                                                                                                                                                                                                                                                                                                                            
## [4] "Keywords = RCC"                                                                                                                                                                                                                                                                                                                                                                              
## [5] "Keywords = carcinoma"                                                                                                                                                                                                                                                                                                                                                                        
## [6] "Keywords = cancer"                                                                                                                                                                                                                                                                                                                                                                           
## [7] "Lot batch = 2004638"                                                                                                                                                                                                                                                                                                                                                                         
## geo_accession 
## [1] "GSM11805"
## last_update_date 
## [1] "May 28 2005"
## molecule_ch1 
## [1] "total RNA"
## organism_ch1 
## [1] "Homo sapiens"
## platform_id 
## [1] "GPL96"
## series_id 
## [1] "GSE781"
## source_name_ch1 
## [1] "Trizol isolation of total RNA from normal tissue adjacent to Renal Cell Carcinoma"
## status 
## [1] "Public on Nov 25 2003"
## submission_date 
## [1] "Oct 20 2003"
## supplementary_file 
## [1] "ftp://ftp.ncbi.nih.gov/pub/geo/DATA/supplementary/samples/GSM11nnn/GSM11805/GSM11805.CEL.gz"
## title 
## [1] "N035 Normal Human Kidney U133A"
## type 
## [1] "RNA"
## An object of class "GEODataTable"
## ****** Column Descriptions ******
##     Column
## 1   ID_REF
## 2    VALUE
## 3 ABS_CALL
##                                                                  Description
## 1                                                                           
## 2                         MAS 5.0 Statistical Algorithm (mean scaled to 500)
## 3 MAS 5.0 Absent, Marginal, Present call  with Alpha1 = 0.05, Alpha2 = 0.065
## ****** Data Table ******
##           ID_REF  VALUE ABS_CALL
## 1 AFFX-BioB-5_at  953.9        P
## 2 AFFX-BioB-M_at 2982.8        P
## 3 AFFX-BioB-3_at 1657.9        P
## 4 AFFX-BioC-5_at 2652.7        P
## 5 AFFX-BioC-3_at 2019.5        P
## 22278 more rows ...</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># and the names of the GPLs represented</span>
<span class="kw">names</span>(<span class="kw"><a href="../reference/Generic%20functions.html">GPLList</a></span>(gse))</code></pre></div>
<pre><code>## [1] "GPL96" "GPL97"</code></pre>
<p><em>See below for an additional, preferred method of obtaining GSE information.</em></p>
</div>
</div>
<div id="converting-to-bioconductor-expressionsets-and-limma-malists" class="section level1">
<h1>Converting to BioConductor ExpressionSets and limma MALists</h1>
<p>GEO datasets are (unlike some of the other GEO entities), quite similar to the <code>limma</code> data structure <code>MAList</code> and to the <code>Biobase</code> data structure <code>ExpressionSet</code>. Therefore, there are two functions, <code>GDS2MA</code> and <code>GDS2eSet</code> that accomplish that task.</p>
<div id="getting-gse-series-matrix-files-as-an-expressionset" class="section level2">
<h2>Getting GSE Series Matrix files as an ExpressionSet</h2>
<p>GEO Series are collections of related experiments. In addition to being available as SOFT format files, which are quite large, NCBI GEO has prepared a simpler format file based on tab-delimited text. The <code>getGEO</code> function can handle this format and will parse very large GSEs quite quickly. The data structure returned from this parsing is a list of ExpressionSets. As an example, we download and parse GSE2553.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Note that GSEMatrix=TRUE is the default</span>
gse2553 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getGEO.html">getGEO</a></span>(<span class="st">'GSE2553'</span>,<span class="dt">GSEMatrix=</span><span class="ot">TRUE</span>)
<span class="kw">show</span>(gse2553)</code></pre></div>
<pre><code>## $GSE2553_series_matrix.txt.gz
## ExpressionSet (storageMode: lockedEnvironment)
## assayData: 12600 features, 181 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: GSM48681 GSM48682 ... GSM48861 (181 total)
##   varLabels: title geo_accession ... data_row_count (30 total)
##   varMetadata: labelDescription
## featureData
##   featureNames: 1 2 ... 12600 (12600 total)
##   fvarLabels: ID PenAt ... Chimeric_Cluster_IDs (13 total)
##   fvarMetadata: Column Description labelDescription
## experimentData: use 'experimentData(object)'
## Annotation: GPL1977</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">show</span>(<span class="kw">pData</span>(<span class="kw">phenoData</span>(gse2553[[<span class="dv">1</span>]]))[<span class="dv">1</span>:<span class="dv">5</span>,<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">6</span>,<span class="dv">8</span>)])</code></pre></div>
<pre><code>##                                                                  title
## GSM48681                      Patient sample ST18, Dermatofibrosarcoma
## GSM48682                           Patient sample ST410, Ewing Sarcoma
## GSM48683                            Patient sample ST130, Sarcoma, NOS
## GSM48684 Patient sample ST293, Malignant Peripheral Nerve Sheath Tumor
## GSM48685                             Patient sample ST367, Liposarcoma
##          type                         source_name_ch1
## GSM48681  RNA                     Dermatofibrosarcoma
## GSM48682  RNA                           Ewing Sarcoma
## GSM48683  RNA                            Sarcoma, NOS
## GSM48684  RNA Malignant Peripheral Nerve Sheath Tumor
## GSM48685  RNA                             Liposarcoma</code></pre>
</div>
<div id="converting-gds-to-an-expressionset" class="section level2">
<h2>Converting GDS to an ExpressionSet</h2>
<p>Taking our <code>gds</code> object from above, we can simply do:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">eset &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Converting.html">GDS2eSet</a></span>(gds,<span class="dt">do.log2=</span><span class="ot">TRUE</span>)</code></pre></div>
<p>Now, <code>eset</code> is an <code>ExpressionSet</code> that contains the same information as in the GEO dataset, including the sample information, which we can see here:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">eset</code></pre></div>
<pre><code>## ExpressionSet (storageMode: lockedEnvironment)
## assayData: 22645 features, 17 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: GSM11815 GSM11832 ... GSM12448 (17 total)
##   varLabels: sample disease.state individual description
##   varMetadata: labelDescription
## featureData
##   featureNames: 200000_s_at 200001_at ... AFFX-TrpnX-M_at (22645
##     total)
##   fvarLabels: ID Gene title ... GO:Component ID (21 total)
##   fvarMetadata: Column labelDescription
## experimentData: use 'experimentData(object)'
##   pubMedIds: 14641932 
## Annotation:</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">pData</span>(eset)[,<span class="dv">1</span>:<span class="dv">3</span>]</code></pre></div>
<pre><code>##            sample disease.state individual
## GSM11815 GSM11815           RCC        035
## GSM11832 GSM11832           RCC        023
## GSM12069 GSM12069           RCC        001
## GSM12083 GSM12083           RCC        005
## GSM12101 GSM12101           RCC        011
## GSM12106 GSM12106           RCC        032
## GSM12274 GSM12274           RCC          2
## GSM12299 GSM12299           RCC          3
## GSM12412 GSM12412           RCC          4
## GSM11810 GSM11810        normal        035
## GSM11827 GSM11827        normal        023
## GSM12078 GSM12078        normal        001
## GSM12099 GSM12099        normal        005
## GSM12269 GSM12269        normal          1
## GSM12287 GSM12287        normal          2
## GSM12301 GSM12301        normal          3
## GSM12448 GSM12448        normal          4</code></pre>
</div>
<div id="converting-gds-to-an-malist" class="section level2">
<h2>Converting GDS to an MAList</h2>
<p>No annotation information (called platform information by GEO) was retrieved from because <code>ExpressionSet</code> does not contain slots for gene information, typically. However, it is easy to obtain this information. First, we need to know what platform this GDS used. Then, another call to <code>getGEO</code> will get us what we need.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#get the platform from the GDS metadata</span>
<span class="kw"><a href="../reference/Generic%20functions.html">Meta</a></span>(gds)$platform</code></pre></div>
<pre><code>## [1] "GPL97"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#So use this information in a call to getGEO</span>
gpl &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getGEO.html">getGEO</a></span>(<span class="dt">filename=</span><span class="kw">system.file</span>(<span class="st">"extdata/GPL97.annot.gz"</span>,<span class="dt">package=</span><span class="st">"GEOquery"</span>))</code></pre></div>
<p>So, <code>gpl</code> now contains the information for GPL5 from GEO. Unlike <code>ExpressionSet</code>, the limma <code>MAList</code> does store gene annotation information, so we can use our newly created <code>gpl</code> of class <code>GPL</code> in a call to <code>GDS2MA</code> like so:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MA &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Converting.html">GDS2MA</a></span>(gds,<span class="dt">GPL=</span>gpl)
<span class="kw">class</span>(MA)</code></pre></div>
<pre><code>## [1] "MAList"
## attr(,"package")
## [1] "limma"</code></pre>
<p>Now, <code>MA</code> is of class <code>MAList</code> and contains not only the data, but the sample information and gene information associated with GDS507.</p>
</div>
<div id="converting-gse-to-an-expressionset" class="section level2">
<h2>Converting GSE to an ExpressionSet</h2>
<p><em>First, make sure that using the method described above in the section ``Getting GSE Series Matrix files as an ExpressionSet&rsquo;&rsquo; for using GSE Series Matrix files is not sufficient for the task, as it is much faster and simpler.</em> If it is not (i.e., other columns from each GSM are needed), then this method will be needed.</p>
<p>Converting a <code>GSE</code> object to an <code>ExpressionSet</code> object currently takes a bit of R data manipulation due to the varied data that can be stored in a <code>GSE</code> and the underlying <code>GSM</code> and <code>GPL</code> objects. However, using a simple example will hopefully be illustrative of the technique.</p>
<p>First, we need to make sure that all of the <code>GSMs</code> are from the same platform:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gsmplatforms &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw"><a href="../reference/Generic%20functions.html">GSMList</a></span>(gse),function(x) {<span class="kw"><a href="../reference/Generic%20functions.html">Meta</a></span>(x)$platform})
<span class="kw">head</span>(gsmplatforms)</code></pre></div>
<pre><code>## $GSM11805
## [1] "GPL96"
## 
## $GSM11810
## [1] "GPL97"
## 
## $GSM11814
## [1] "GPL96"
## 
## $GSM11815
## [1] "GPL97"
## 
## $GSM11823
## [1] "GPL96"
## 
## $GSM11827
## [1] "GPL97"</code></pre>
<p>Indeed, there are two GPLs, GPL96 and GPL97, as their platforms (which we could have determined by looking at the GPLList for <code>gse</code>). We can filter the original GSMList to include only those GSMs with the GPL96 platform and use this list for further processing</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gsmlist =<span class="st"> </span><span class="kw">Filter</span>(function(gsm) {<span class="kw"><a href="../reference/Generic%20functions.html">Meta</a></span>(gsm)$platform==<span class="st">'GPL96'</span>},<span class="kw"><a href="../reference/Generic%20functions.html">GSMList</a></span>(gse))
<span class="kw">length</span>(gsmlist)</code></pre></div>
<pre><code>## [1] 17</code></pre>
<p>So, now we would like to know what column represents the data that we would like to extract. Looking at the first few rows of the Table of a single GSM will likely give us an idea (and by the way, GEO uses a convention that the column that contains the single measurement for each array is called the <code>VALUE</code> column, which we could use if we don&rsquo;t know what other column is most relevant).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/Generic%20functions.html">Table</a></span>(gsmlist[[<span class="dv">1</span>]])[<span class="dv">1</span>:<span class="dv">5</span>,]</code></pre></div>
<pre><code>##           ID_REF  VALUE ABS_CALL
## 1 AFFX-BioB-5_at  953.9        P
## 2 AFFX-BioB-M_at 2982.8        P
## 3 AFFX-BioB-3_at 1657.9        P
## 4 AFFX-BioC-5_at 2652.7        P
## 5 AFFX-BioC-3_at 2019.5        P</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># and get the column descriptions</span>
<span class="kw"><a href="../reference/Generic%20functions.html">Columns</a></span>(gsmlist[[<span class="dv">1</span>]])[<span class="dv">1</span>:<span class="dv">5</span>,]</code></pre></div>
<pre><code>##        Column
## 1      ID_REF
## 2       VALUE
## 3    ABS_CALL
## NA       &lt;NA&gt;
## NA.1     &lt;NA&gt;
##                                                                     Description
## 1                                                                              
## 2                            MAS 5.0 Statistical Algorithm (mean scaled to 500)
## 3    MAS 5.0 Absent, Marginal, Present call  with Alpha1 = 0.05, Alpha2 = 0.065
## NA                                                                         &lt;NA&gt;
## NA.1                                                                       &lt;NA&gt;</code></pre>
<p>We will indeed use the <code>VALUE</code> column. We then want to make a matrix of these values like so:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get the probeset ordering</span>
probesets &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Generic%20functions.html">Table</a></span>(<span class="kw"><a href="../reference/Generic%20functions.html">GPLList</a></span>(gse)[[<span class="dv">1</span>]])$ID
<span class="co"># make the data matrix from the VALUE columns from each GSM</span>
<span class="co"># being careful to match the order of the probesets in the platform</span>
<span class="co"># with those in the GSMs</span>
data.matrix &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="st">'cbind'</span>,<span class="kw">lapply</span>(gsmlist,function(x) 
                                      {tab &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Generic%20functions.html">Table</a></span>(x)
                                       mymatch &lt;-<span class="st"> </span><span class="kw">match</span>(probesets,tab$ID_REF)
                                       <span class="kw">return</span>(tab$VALUE[mymatch])
                                     }))
data.matrix &lt;-<span class="st"> </span><span class="kw">apply</span>(data.matrix,<span class="dv">2</span>,function(x) {<span class="kw">as.numeric</span>(<span class="kw">as.character</span>(x))})
data.matrix &lt;-<span class="st"> </span><span class="kw">log2</span>(data.matrix)
data.matrix[<span class="dv">1</span>:<span class="dv">5</span>,]</code></pre></div>
<pre><code>##       GSM11805  GSM11814  GSM11823  GSM11830  GSM12067  GSM12075  GSM12079
## [1,] 10.926963 11.105254 11.275019 11.438636 11.424376 11.222795 11.469845
## [2,]  5.749534  7.908092  7.093814  7.514122  7.901470  6.407693  5.165912
## [3,]  7.066089  7.750205  7.244126  7.962896  7.337176  6.569856  7.477354
## [4,] 12.660353 12.479755 12.215897 11.458355 11.397568 12.529870 12.240046
## [5,]  6.195741  6.061776  6.565293  6.583459  6.877744  6.652486  3.981853
##       GSM12098  GSM12100  GSM12105  GSM12268  GSM12270  GSM12283  GSM12298
## [1,] 10.823367 10.835971 10.810893 11.062653 10.323055 11.181028 11.566387
## [2,]  6.556123  8.207014  6.816344  6.563768  7.353147  5.770829  6.912889
## [3,]  7.708739  7.428779  7.754888  7.126188  8.742815  7.339850  7.602142
## [4,] 12.336534 11.762839 11.237509 12.412490 11.213408 12.678380 12.232901
## [5,]  5.501439  6.247928  6.017922  6.525129  6.683696  5.918863  5.837943
##       GSM12300  GSM12399  GSM12444
## [1,] 11.078151 11.535178 11.105450
## [2,]  4.812498  7.471675  7.488644
## [3,]  7.383704  7.432959  7.381110
## [4,] 12.090939 11.421802 12.172834
## [5,]  6.281698  5.419539  5.469235</code></pre>
<p>Note that we do a <code>match</code> to make sure that the values and the platform information are in the same order. Finally, to make the <code>ExpressionSet</code> object:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(Biobase)
<span class="co"># go through the necessary steps to make a compliant ExpressionSet</span>
<span class="kw">rownames</span>(data.matrix) &lt;-<span class="st"> </span>probesets
<span class="kw">colnames</span>(data.matrix) &lt;-<span class="st"> </span><span class="kw">names</span>(gsmlist)
pdata &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">samples=</span><span class="kw">names</span>(gsmlist))
<span class="kw">rownames</span>(pdata) &lt;-<span class="st"> </span><span class="kw">names</span>(gsmlist)
pheno &lt;-<span class="st"> </span><span class="kw">as</span>(pdata,<span class="st">"AnnotatedDataFrame"</span>)
eset2 &lt;-<span class="st"> </span><span class="kw">new</span>(<span class="st">'ExpressionSet'</span>,<span class="dt">exprs=</span>data.matrix,<span class="dt">phenoData=</span>pheno)
eset2</code></pre></div>
<pre><code>## ExpressionSet (storageMode: lockedEnvironment)
## assayData: 22283 features, 17 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: GSM11805 GSM11814 ... GSM12444 (17 total)
##   varLabels: samples
##   varMetadata: labelDescription
## featureData: none
## experimentData: use 'experimentData(object)'
## Annotation:</code></pre>
<p>So, using a combination of <code>lapply</code> on the GSMList, one can extract as many columns of interest as necessary to build the data structure of choice. Because the GSM data from the GEO website are fully downloaded and included in the <code>GSE</code> object, one can extract foreground and background as well as quality for two-channel arrays, for example. Getting array annotation is also a bit more complicated, but by replacing ``platform&rsquo;&rsquo; in the lapply call to get platform information for each array, one can get other information associated with each array.</p>
</div>
</div>
<div id="accessing-raw-data-from-geo" class="section level1">
<h1>Accessing Raw Data from GEO</h1>
<p>NCBI GEO accepts (but has not always required) raw data such as .CEL files, .CDF files, images, etc. Sometimes, it is useful to get quick access to such data. A single function, <code>getGEOSuppFiles</code>, can take as an argument a GEO accession and will download all the raw data associate with that accession. By default, the function will create a directory in the current working directory to store the raw data for the chosen GEO accession. Combining a simple <code>sapply</code> statement or other loop structure with <code>getGEOSuppFiles</code> makes for a very simple way to get gobs of raw data quickly and easily without needing to know the specifics of GEO raw data URLs.</p>
</div>
<div id="use-cases" class="section level1">
<h1>Use Cases</h1>
<p>GEOquery can be quite powerful for gathering a lot of data quickly. A few examples can be useful to show how this might be done for data mining purposes.</p>
<div id="getting-all-series-records-for-a-given-platform" class="section level2">
<h2>Getting all Series Records for a Given Platform</h2>
<p>For data mining purposes, it is sometimes useful to be able to pull all the GSE records for a given platform. GEOquery makes this very easy, but a little bit of knowledge of the GPL record is necessary to get started. The GPL record contains both the GSE and GSM accessions that reference it. Some code is useful to illustrate the point:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gpl97 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getGEO.html">getGEO</a></span>(<span class="st">'GPL97'</span>)
<span class="kw"><a href="../reference/Generic%20functions.html">Meta</a></span>(gpl97)$title</code></pre></div>
<pre><code>## [1] "[HG-U133B] Affymetrix Human Genome U133B Array"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw"><a href="../reference/Generic%20functions.html">Meta</a></span>(gpl97)$series_id)</code></pre></div>
<pre><code>## [1] "GSE362" "GSE473" "GSE620" "GSE674" "GSE781" "GSE907"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(<span class="kw"><a href="../reference/Generic%20functions.html">Meta</a></span>(gpl97)$series_id)</code></pre></div>
<pre><code>## [1] 158</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw"><a href="../reference/Generic%20functions.html">Meta</a></span>(gpl97)$sample_id)</code></pre></div>
<pre><code>## [1] "GSM3922" "GSM3924" "GSM3926" "GSM3928" "GSM3930" "GSM3932"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(<span class="kw"><a href="../reference/Generic%20functions.html">Meta</a></span>(gpl97)$sample_id)</code></pre></div>
<pre><code>## [1] 7701</code></pre>
<p>The code above loads the GPL97 record into R. The Meta method extracts a list of header information from the GPL record. The <code>title</code> gives the human name of the platform. The <code>series_id</code> gives a vector of series ids. Note that there are 158 series associated with this platform and 7701 samples. Code like the following could be used to download all the samples or series. I show only the first 5 samples as an example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gsmids &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Generic%20functions.html">Meta</a></span>(gpl97)$sample_id
gsmlist &lt;-<span class="st"> </span><span class="kw">sapply</span>(gsmids[<span class="dv">1</span>:<span class="dv">5</span>],getGEO)
<span class="kw">names</span>(gsmlist)</code></pre></div>
<pre><code>## [1] "GSM3922" "GSM3924" "GSM3926" "GSM3928" "GSM3930"</code></pre>
</div>
</div>
<div id="conclusion" class="section level1">
<h1>Conclusion</h1>
<p>The GEOquery package provides a bridge to the vast array resources contained in the NCBI GEO repositories. By maintaining the full richness of the GEO data rather than focusing on getting only the ``numbers&rsquo;&rsquo;, it is possible to integrate GEO data into current Bioconductor data structures and to perform analyses on that data quite quickly and easily. These tools will hopefully open GEO data more fully to the array community at large.</p>
<div id="citing-geoquery" class="section level2">
<h2>Citing GEOquery</h2>
<p>Please consider citing GEOquery if used in support of your own research:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">citation</span>(<span class="st">"GEOquery"</span>)</code></pre></div>
<pre><code>## 
## To cite package 'GEOquery' in publications use:
## 
##   Sean Davis (2016). GEOquery: Get data from NCBI Gene Expression
##   Omnibus (GEO). R package version 2.39.4.
##   https://github.com/seandavi/GEOquery
## 
## A BibTeX entry for LaTeX users is
## 
##   @Manual{,
##     title = {GEOquery: Get data from NCBI Gene Expression Omnibus (GEO)},
##     author = {Sean Davis},
##     year = {2016},
##     note = {R package version 2.39.4},
##     url = {https://github.com/seandavi/GEOquery},
##   }
## 
## ATTENTION: This citation information has been auto-generated from
## the package DESCRIPTION file and may need manual editing, see
## 'help("citation")'.</code></pre>
</div>
<div id="reporting-problems-or-bugs" class="section level2">
<h2>Reporting problems or bugs</h2>
<p>If you run into problems using GEOquery, the <a href="https://support.bioconductor.org/">Bioconductor Support site</a> is a good first place to ask for help. If you are convinced that there is a bug in GEOquery (this is pretty unusual, but not unheard of), feel free to submit an issue on the <a href="https://github.com/seandavi/GEOquery">GEOquery github site</a> or file a bug report directly from R (will open a new github issue):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">bug.report</span>(<span class="dt">package=</span><span class="st">'GEOquery'</span>)</code></pre></div>
</div>
</div>
<div id="session-info" class="section level1">
<h1>Session info</h1>
<p>The following package and versions were used in the production of this vignette.</p>
<pre><code>## R version 3.3.1 (2016-06-21)
## Platform: x86_64-apple-darwin13.4.0 (64-bit)
## Running under: OS X 10.11.6 (El Capitan)
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] parallel  stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
## [1] limma_3.29.20       knitr_1.14          GEOquery_2.39.4    
## [4] Biobase_2.33.2      BiocGenerics_0.19.2
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.7        roxygen2_5.0.1     xml2_1.0.0.9001   
##  [4] whisker_0.3-2      magrittr_1.5       devtools_1.12.0   
##  [7] R6_2.2.0           stringr_1.1.0      httr_1.2.1        
## [10] tcltk_3.3.1        tools_3.3.1        git2r_0.15.0      
## [13] withr_1.0.2        htmltools_0.3.5    yaml_2.1.13       
## [16] lazyeval_0.2.0     assertthat_0.1     digest_0.6.10     
## [19] rprojroot_1.0-2    tibble_1.2         pkgdown_0.1.0.9000
## [22] crayon_1.3.2       formatR_1.4        purrr_0.2.2       
## [25] codetools_0.2-14   bitops_1.0-6       RCurl_1.95-4.8    
## [28] curl_2.1           memoise_1.0.0      evaluate_0.10     
## [31] rmarkdown_1.1.9005 stringi_1.1.1      desc_1.0.1        
## [34] highlight_0.4.7    XML_3.98-1.4</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs">
    <div id="sidebar">
            <div id="tocnav">
        <h2>Contents</h2>
        <ul class="nav nav-pills nav-stacked"><li><a href="#overview-of-geo">Overview of GEO</a><ul class="nav nav-pills nav-stacked"><li><a href="#platforms">Platforms</a></li>
        <li><a href="#samples">Samples</a></li>
        <li><a href="#series">Series</a></li>
        <li><a href="#datasets">Datasets</a></li>
        </ul></li>
        <li><a href="#getting-started-using-geoquery">Getting Started using GEOquery</a></li>
        <li><a href="#geoquery-data-structures">GEOquery Data Structures</a><ul class="nav nav-pills nav-stacked"><li><a href="#the-gds-gsm-and-gpl-classes">The GDS, GSM, and GPL classes</a></li>
        <li><a href="#the-gse-class">The GSE class</a></li>
        </ul></li>
        <li><a href="#converting-to-bioconductor-expressionsets-and-limma-malists">Converting to BioConductor ExpressionSets and limma MALists</a><ul class="nav nav-pills nav-stacked"><li><a href="#getting-gse-series-matrix-files-as-an-expressionset">Getting GSE Series Matrix files as an ExpressionSet</a></li>
        <li><a href="#converting-gds-to-an-expressionset">Converting GDS to an ExpressionSet</a></li>
        <li><a href="#converting-gds-to-an-malist">Converting GDS to an MAList</a></li>
        <li><a href="#converting-gse-to-an-expressionset">Converting GSE to an ExpressionSet</a></li>
        </ul></li>
        <li><a href="#accessing-raw-data-from-geo">Accessing Raw Data from GEO</a></li>
        <li><a href="#use-cases">Use Cases</a><ul class="nav nav-pills nav-stacked"><li><a href="#getting-all-series-records-for-a-given-platform">Getting all Series Records for a Given Platform</a></li>
        </ul></li>
        <li><a href="#conclusion">Conclusion</a><ul class="nav nav-pills nav-stacked"><li><a href="#citing-geoquery">Citing GEOquery</a></li>
        <li><a href="#reporting-problems-or-bugs">Reporting problems or bugs</a></li>
        </ul></li>
        <li><a href="#session-info">Session info</a></li>
        </ul></div>
          </div>
  </div>

</div>


      <footer><p>Built by <a href="http://hadley.github.io/pkgdown/">pkgdown</a>. Styled with <a href="http://getbootstrap.com">Bootstrap 3</a>.</p>
      </footer></div>

  </body></html>
